### ISSUE TITLE
iOS private mode: Missing check for tus.isSupported.

### ISSUE THREAD

**Issue by mmathys** [tus/tus-js-client#49]
I recently encountered an issue where the value of `tus.isSupported` was set to `true` but the browser didn't support tus:

The private mode of Safari on iOS doesn't allow access to the local storage ([see this thread](http://stackoverflow.com/questions/21159301/quotaexceedederror-dom-exception-22-an-attempt-was-made-to-add-something-to-st)).

Therefore it might be good to implement a check which checks whether the local storage is accessible, [something like this maybe](http://stackoverflow.com/questions/16427636/check-if-localstorage-is-available).

**Comment by Acconut** [comment]
Intentionally, `isSupported` may be true even if local storage is not available. The reason is that it should indicate whether tus-js-client can upload at all which requires file slicing and support for XHR but no Web Storage. If you are interested in whether it is able to store URLs in the local storage, you have to examine `canStoreURLs` (see https://github.com/tus/tus-js-client#tuscanstoreurls).

**Comment by mmathys** [comment]
I see, I wasn't aware of `canStoreURLs`.

However I'm not sure the private mode on Safari on iOS really sets the `canStoreURLs` to false: if I'm not mistaken, [tus is only checking whether the local storage exists, but not whether it's accepting any values](https://github.com/tus/tus-js-client/blob/e7a07af900c058a68fdad04001a9cde21a29d0c7/lib/browser/storage.js#L12).

Local storage _exists_ in private mode, but it's not accepting any values and throws a `QuotaExceededError` when attempting to actually save values, [see this thread](http://stackoverflow.com/questions/21159301/quotaexceedederror-dom-exception-22-an-attempt-was-made-to-add-something-to-st). It's a strange behaviour.

**Comment by mmathys** [comment]
The modernizr's approach handles this issue.

```javascript
function lsTest () {
  var test = 'test';
  try {
    localStorage.setItem(test, test);
    localStorage.removeItem(test);
    return true;
  } catch(e) {
    return false;
  }
}

if (lsTest () === true) {
    // available
} else {
    // unavailable
}
```

**Comment by Acconut** [comment]
> However I'm not sure the private mode on Safari on iOS really sets the canStoreURLs to false: if I'm not mistaken, tus is only checking whether the local storage exists, but not whether it's accepting any values.

In the past, this check has served us well in detecting whether local storage is usable, as accessing the length of it is usually not permitted if the entire Web Storage is unavailable. However, this may not be the case in Safari's private mode. I will take a look at it but this may take some time as I do not have an iOS device at hand.

**Comment by mmathys** [comment]
@Acconut If you like you can assign this issue to me, I'm working on a project right now which involves tus and iOS devices. As far as my company is concerned, working on tus is considered as working on the product.

**Comment by Acconut** [comment]
@mmathys That would be very helpful if you could look into this. Thank to you and your company already in advance.

**Comment by Acconut** [comment]
Closing in favor of #53.

**Comment by Acconut** [comment]
The according patch is available in `v1.4.2`.
